trapper/commands.py

```diff
@@ -94,20 +104,172 @@ def add_subparser(
         return subparser
 
 
+@Subcommand.register("run-distributed")
+class DistributedRun(Subcommand):
+    """Start an experiment in a distributed data parallel fashion using the config file.
+    It utilizes pytorch's elastic run under the hood. To see supported features see
+    https://pytorch.org/docs/stable/elastic/run.html
+
+    Usage:
+        Basic:
+            ` trapper run-distributed experiment.jsonnet <--other-kwargs vals>`
+    """
+
+    def add_subparser(
+        self, parser: argparse._SubParsersAction
+    ) -> argparse.ArgumentParser:
+        description = """Run the experiment specified in the config file in a 
+            distributed fashion across multiple devices (DDP)."""
+        subparser = parser.add_parser(
+            self.name,
+            description=description,
+            help="Start distributed training of a model.",
+        )
+
+        subparser.add_argument(
+            "config_path",
+            type=str,
+            help="path to the experiment config file in `json` or `jsonnet`"
+            " format describing the model, dataset and other details.",
+        )
+
+        subparser.add_argument(
+            "-o",
+            "--overrides",
+            type=str,
+            default="",
+            help=(
+                "a json or jsonnet structure used to override the experiment "
+                "configuration, e.g., '{\"optimizer.lr\": 1e-5}'.  Nested "
+                "parameters can be specified either with nested dictionaries "
+                "or with dot syntax."
+            ),
+        )
+
+        subparser.set_defaults(func=run_distributed)
+
+        return subparser
+
+
+@record
+def run_distributed(args=None):
+    """
+    Taken from
+    https://github.com/pytorch/pytorch/blob/4618371da56c887195e2e1d16dad2b9686302800/torch/distributed/run.py
+    and modified to properly run by trapper.
+    """
+
+    def merge_json_str(opts: List[str]) -> None:
+        """
+        Merges the JSON structures from given command line arguments.
+        Implements a simple stack to merge json. Note that it modifies
+        given args list inplace.
+
+        Args:
+            *opts:
+
+        Returns:
+            List of strings json where JSON structures merged.
+        """
+        start = step = 1
+
+        i = 0
+        while i < len(opts):
+            if opts[i].count("{") >= 1:
+                start = i
+                break
+            i += 1
+
+        while start + step < len(opts) + 1:
+            end = start + step
+            try:
+                json_shards = " ".join(opts[start:end])
+                json.loads(json_shards)
+            except json.decoder.JSONDecodeError:
+                step += 1
+            else:
+                json_str = " ".join(opts[start:end])
+                del opts[start:end]
+                opts.insert(start, json_str)
+                break
```

Comment:
I think this logic assumes there is a single json structure supplied (if supplied at all).  This means, multiple independent json arguments are invalid and-or unexpected etc. Can you add an explanation on this?

Reply:
@cemilcengiz 

> I think this logic assumes there is a single json structure supplied (if supplied at all).

Yes it assumes that there will be only a single json structure in commandline which is given by flag `--overrides`, it does not accept a list anyway, so the assumption is valid in the current use case.

> This means, multiple independent json arguments are invalid and-or unexpected etc.

Yes, they will be invalid and this is intentional.

> Can you add an explanation on this?

I included a docstring with the function, what additional info do you mean to elaborate ?

Reply:
Well, you can add that it is used to parse the `--overrides` arguments maybe, it felt cryptic to me. 

Related to this: The docstring has the statement of `Merges the JSON structures from given command line arguments.` which is not accurate as it parses only a single json argument. Maybe, you can say it is used to parse possible `--overrides` argument.